include/flux/op/zip.hpp

```diff
@@ -169,6 +127,99 @@ struct sequence_traits<detail::zip_adaptor<Bases...>>
             return std::min({flux::size(args)...});
         }, self.bases_);
     }
+};
+
+
+namespace detail {
+
+template <sequence... Bases>
+struct zip_adaptor : inline_sequence_base<zip_adaptor<Bases...>> {
+private:
+    pair_or_tuple_t<Bases...> bases_;
+
+    friend struct sequence_traits<zip_adaptor>;
+    friend struct zip_traits_base<Bases...>;
+
+public:
+    constexpr explicit zip_adaptor(decays_to<Bases> auto&&... bases)
+        : bases_(FLUX_FWD(bases)...)
+    {}
+};
+
+struct zip_fn {
+    template <adaptable_sequence... Seqs>
+    [[nodiscard]]
+    constexpr auto operator()(Seqs&&... seqs) const
+    {
+        if constexpr (sizeof...(Seqs) == 0) {
+            return empty<std::tuple<>>;
+        } else {
+            return zip_adaptor<std::decay_t<Seqs>...>(FLUX_FWD(seqs)...);
+        }
+    }
+};
+
+template <typename Func, sequence... Bases>
+struct zip_map_adaptor : inline_sequence_base<zip_map_adaptor<Func, Bases...>> {
+private:
+    pair_or_tuple_t<Bases...> bases_;
+    Func func_;
+
+    friend struct sequence_traits<zip_map_adaptor>;
+    friend struct zip_traits_base<Bases...>;
+
+public:
+    constexpr explicit zip_map_adaptor(Func&& func, decays_to<Bases> auto&&... bases)
+        : bases_(FLUX_FWD(bases)...), func_(std::move(func))
+    {}
+};
+
+struct zip_map_fn {
+    template <typename Func, adaptable_sequence... Seqs>
+        requires std::is_invocable_v<Func, element_t<Seqs>...>
+    [[nodiscard]]
+    constexpr auto operator()(Func func, Seqs&&... seqs) const
+    {
+        if constexpr (sizeof...(Seqs) == 0) {
+            return empty<std::invoke_result_t<Func>>;
```

Comment:
At the moment, `flux::empty<T>` is constrained to require `std::is_object_v<T>`, so this will be a problem if `func()` returns a reference type. 

I have to confess that I can't remember the reason for having that constraint on `empty`! It doesn't seem like it would be problematic to remove it, so that we can make this work.